File: PermisosJava/app/src/main/java/com/example/permisosjava/MainActivity.java
package com.example.permisosjava;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {
    private static final int RECORD_REQUEST_CODE=500;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        solicitarPermiso();
    }
    private void solicitarPermiso(){
        if(ContextCompat.checkSelfPermission(this,Manifest.permission.RECORD_AUDIO)!= PackageManager.PERMISSION_GRANTED){
            Toast.makeText(this, "Permiso para accder al microfono es requerido para esta aplicacion para grabar audio", Toast.LENGTH_LONG).show();
            
            if(ActivityCompat.shouldShowRequestPermissionRationale(this,Manifest.permission.RECORD_AUDIO)){
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.RECORD_AUDIO},RECORD_REQUEST_CODE);
            }else{
                ActivityCompat.requestPermissions(this,new String[]{Manifest.permission.RECORD_AUDIO},RECORD_REQUEST_CODE);
            }
        } else if (ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO)==PackageManager.PERMISSION_GRANTED) {
            
        }
    }
    public void onRequestPermissionsResult(int requestCode, String permissions[],int[] grantResults){
        super.onRequestPermissionsResult(requestCode,permissions,grantResults);
        switch (requestCode){
            case RECORD_REQUEST_CODE:{
                if(grantResults.length>0 && grantResults[0]==PackageManager.PERMISSION_GRANTED){
                    Toast.makeText(this, "Permiso Concedido para grabar audio", Toast.LENGTH_SHORT).show();
                }else{
                    Toast.makeText(this, "Permiso Denegado para grabar audio", Toast.LENGTH_SHORT).show();
                }
            }
        }
    }
}

File: Permisos/app/src/main/java/com/example/permisos/MainActivity.kt
package com.example.permisos

import android.content.pm.PackageManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.Manifest
import android.widget.Toast
import android.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

private val TAG="Permisos Practica 4"
private val  CODIGOSOLICITUDGRABAR=101
class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        configurarPermisos()
    }
    private fun configurarPermisos(){
        val permisos=ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO)
        if(permisos!=PackageManager.PERMISSION_GRANTED){
            Log.i(TAG,"Permiso par agrabar denegado")
            if(ActivityCompat.shouldShowRequestPermissionRationale(this,android.Manifest.permission.RECORD_AUDIO)){
                val builder =AlertDialog.Builder(this)
                builder.setMessage("Permisos para acceder el microfono es requerido para grabar audio.").setTitle("Permission Requeried")
                builder.setPositiveButton("Ok"){dialog,id->Log.i(TAG, "Seleccionado")
                    hacerSolicitud()
                }
                val dialog=builder.create()
                dialog.show()
            }else{
                hacerSolicitud()
            }
        }
    }
    private fun hacerSolicitud(){
        ActivityCompat.requestPermissions(this, arrayOf(android.Manifest.permission.RECORD_AUDIO),
            CODIGOSOLICITUDGRABAR)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when(requestCode){
            CODIGOSOLICITUDGRABAR->{
                if(grantResults.isEmpty()||grantResults[0]!=PackageManager.PERMISSION_GRANTED){
                    Log.i(TAG,"Permisos a sido denegado por el usuario")
                    Toast.makeText(applicationContext, "Permiso a sido denegado por el usuario", Toast.LENGTH_SHORT).show()
                }else{
                    Log.i(TAG,"Permisos a sido concedido por el usuario")
                    Toast.makeText(applicationContext, "Permiso a sido concedido por el usuario", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }
}